### PR TITLE
docs: Move cve_scan and generation of external dependencies rst to bazel

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: docs: Move generation of external dependencies rst to bazel
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: tools/dependency/utils.py [hunk 1]
```diff
@@ -14,12 +15,14 @@ def LoadModule(name, path):
   return module
 
 
+API_PATH = os.getenv("API_PATH", "external/envoy_api_canonical")
```
Comment (phlax) [inline tools/dependency/utils.py:18]: `"api/"`

File: tools/extensions/BUILD [hunk 2]
```diff
@@ -21,5 +21,4 @@ py_binary(
     name = "generate_extension_rst",
     srcs = ["generate_extension_rst.py"],
     data = [":generate_extension_db"],
-    visibility = ["//visibility:public"],
```
Comment (phlax) [inline tools/extensions/BUILD:24]: ?

File: tools/dependency/BUILD [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+licenses(["notice"])  # Apache 2
+
+envoy_package()
+
+py_binary(
+    name = "utils",
+    srcs = ["utils.py"],
+    data = [
+        "//bazel:repository_locations.bzl",
+        "@com_google_protobuf//:protobuf_python",
```
Comment (phlax) [inline tools/dependency/BUILD:13]: ?

File: tools/dependency/utils.py (hunk 1 above)
Comment (phlax) [reply tools/dependency/utils.py:18]: actuall this is the only place it is used so the old default doesnt matter

```
$ git grep dependency | grep utils
tools/dependency/generate_external_dep_rst.py:from tools.dependency import utils as dep_utils
```

File: docs/build.sh [hunk 4]
```diff
@@ -52,7 +52,6 @@ rm -rf "${DOCS_OUTPUT_DIR}"
 mkdir -p "${DOCS_OUTPUT_DIR}"
 
 rm -rf "${GENERATED_RST_DIR}"
-mkdir -p "${GENERATED_RST_DIR}"
```
Comment (phlax) [inline docs/build.sh:55]: leave this for now

File: tools/dependency/cve_scan.py [hunk 5]
```diff
@@ -7,11 +7,14 @@
 import datetime as dt
 import gzip
 import json
+import os
 import re
 import sys
 import textwrap
 import urllib.request
 
+os.environ["API_PATH"] = "api"
```
Comment (phlax) [inline tools/dependency/cve_scan.py:16]: add comment here

File: tools/dependency/utils.py (hunk 1 above)
Comment (phlax) [reply tools/dependency/utils.py:18]: actually not - that was incorrect - the cve_scan module imports it...

**Comment by phlax** [comment]
generated rst is here:

https://storage.googleapis.com/envoy-pr/bbaf31b/docs/intro/arch_overview/security/external_deps.html

**Comment by ggreenway** [comment]
I'm not the right reviewer for this; I don't know bazel well enough. Maybe @lizan ?

**Comment by phlax** [comment]
cc @htuch @moderation

**Comment by phlax** [comment]
thanks @ggreenway

**Comment by moderation** [comment]
This LGTM but definitely needs @htuch with his combination of Bazel and Python skills to review

File: tools/dependency/BUILD [hunk 6]
```diff
@@ -0,0 +1,26 @@
+load("@rules_python//python:defs.bzl", "py_binary")
+load("//bazel:envoy_build_system.bzl", "envoy_package")
+
+licenses(["notice"])  # Apache 2
+
+envoy_package()
```
Comment (phlax) [inline tools/dependency/BUILD:6]: not sure if this should be an `envoy_package()`

Comment (htuch) [reply tools/dependency/BUILD:6]: This is fine, it's just setting visibility (you can skip it on the targets).

File: tools/dependency/cve_scan.py [hunk 7]
```diff
@@ -7,6 +7,7 @@
 import datetime as dt
 import gzip
 import json
+import os
```
Comment (htuch) [inline tools/dependency/cve_scan.py:10]: Seems unrelated and that it should have been broken before if not there.

File: tools/dependency/generate_external_dep_rst.py [hunk 8]
```diff
@@ -3,10 +3,16 @@
...
 from collections import defaultdict, namedtuple
+import os
 import pathlib
 import sys
 import urllib.parse
 
+# TODO(phlax): remove this
+# this needs to be before the utils import
+# if we move cve_scan to a bazel job then we can remove
+os.environ["API_PATH"] = "external/envoy_api_canonical"
```
Comment (htuch) [inline tools/dependency/generate_external_dep_rst.py:14]: Oh, I see. How complex is it to make Bazelify cve_scan?

**Review by htuch — COMMENTED** [review]
Looks good, but would be idea if we didn't need any special casing on the border of Bazel and non-Bazel Python use..

File: tools/dependency/generate_external_dep_rst.py (hunk 8 above)
Comment (phlax) [reply tools/dependency/generate_external_dep_rst.py:14]: not too complex i think

would you prefer to update that first ?

File: tools/dependency/BUILD (hunk 6 above)
Comment (phlax) [reply tools/dependency/BUILD:6]: targets updated

File: tools/dependency/cve_scan.py (hunk 7 above)
Comment (phlax) [reply tools/dependency/cve_scan.py]: removed

File: tools/dependency/generate_external_dep_rst.py (hunk 8 above)
Comment (phlax) [reply tools/dependency/generate_external_dep_rst.py:14]: pr added for `cve_scan` here #15342

**Comment by phlax** [comment]
one hesitation i have with this PR is that i think it would be better for the tools in `/tools/extension` and `/tools/dependency` to all just generate eg json - and then have that all put together into rst in a doc tool

i think we can do that once we have finished moving the docs generation to bazel tho

File: ci/run_envoy_docker.sh [hunk 9]
```diff
@@ -44,6 +44,8 @@ else
     && useradd -o --uid $(id -u) --gid $(id -g) --no-create-home --home-dir /build envoybuild \
     && usermod -a -G pcap envoybuild \
     && chown envoybuild:envoygroup /build \
+    && update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.8 1 \
+    && pip3 install -U virtualenv \
     && sudo -EHs -u envoybuild bash -c 'cd /source && $*'")
```
Comment (phlax) [inline ci/run_envoy_docker.sh:49]: these wont be necessary once #15358 lands

File: tools/dependency/generate_external_dep_rst.py (hunk 8 above)
Comment (phlax) [reply tools/dependency/generate_external_dep_rst.py:14]: and merged with this PR to simplify...

**Comment by htuch** [comment]
@phlax JSON (or proto) for intermediate artifacts makes sense to me.

**Review by htuch — COMMENTED** [review]
LGTM for draft, please ping when ready for full review.
/wait

**Comment by phlax** [comment]
@htuch ping - should be ready for final review (nothing changed since last review)

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by htuch** [comment]
@phlax CI seems stuck, can you merge main and push?

**Comment by moderation** [comment]
@phlax I run a pretty heavily modified version of `tools/dependency/release_dates.py` multiple times a day as part of deps shepharding. I think this PR introduced a breakage.
```
python3.9 tools/dependency/release_dates.py api/bazel/repository_locations.bzl | rg -vi github.re
Traceback (most recent call last):
  File "/home/moderation/Library/envoyproxy/envoy/tools/dependency/release_dates.py", line 156, in <module>
    spec_loader = utils.repository_locations_utils.load_repository_locations_spec
AttributeError: module 'utils' has no attribute 'repository_locations_utils'
```
Should I be running this now via `bazel`?

**Comment by moderation** [comment]
@mattklein123 was seeing CI failures and we discussed in Slack. The PR has been reverted to unblock CI - https://github.com/envoyproxy/envoy/pull/15421

**Pull request merged**